larray/inout/pandas.py

```diff
@@ -39,61 +39,113 @@ def parse(s):
         return s
 
 
-def df_labels(df, sort=True):
+def index_to_labels(idx, sort=True):
     """
     Returns unique labels for each dimension.
     """
-    idx = df.index
     if isinstance(idx, pd.core.index.MultiIndex):
         if sort:
             return list(idx.levels)
         else:
             return [list(unique(idx.get_level_values(l))) for l in idx.names]
     else:
         assert isinstance(idx, pd.core.index.Index)
-        # use .values if needed
-        return [idx]
+        labels = list(idx.values)
+        return [sorted(labels) if sort else labels]
 
 
-def cartesian_product_df(df, sort_rows=False, sort_columns=False, **kwargs):
-    labels = df_labels(df, sort=sort_rows)
-    if sort_rows:
-        new_index = pd.MultiIndex.from_product(labels)
+def cartesian_product_df(df, sort_rows=False, sort_columns=False, fill_value=nan, **kwargs):
+    idx = df.index
+    labels = index_to_labels(idx, sort=sort_rows)
+    if isinstance(idx, pd.core.index.MultiIndex):
+        if sort_rows:
+            new_index = pd.MultiIndex.from_product(labels)
+        else:
+            new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
     else:
-        new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
+        if sort_rows:
+            new_index = pd.Index(labels[0], name=idx.name)
+        else:
+            new_index = idx
     columns = sorted(df.columns) if sort_columns else list(df.columns)
     # the prodlen test is meant to avoid the more expensive array_equal test
     prodlen = np.prod([len(axis_labels) for axis_labels in labels])
     if prodlen == len(df) and columns == list(df.columns) and np.array_equal(df.index.values, new_index.values):
         return df, labels
-    return df.reindex(new_index, columns, **kwargs), labels
+    return df.reindex(index=new_index, columns=columns, fill_value=fill_value, **kwargs), labels
 
 
-def from_series(s, sort_rows=False):
-    """
-    Converts Pandas Series into 1D LArray.
+def from_series(s, sort_rows=False, fill_value=nan, meta=None, **kwargs):
```

Comment:
the support for multiIndex in from_series needs to be in the release notes I think. fill_value and meta can be omitted from them IMO.